sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -4483,6 +4478,17 @@ class Dataset[T] private[sql](
     }
   }
 
+  /** Returns a optimized plan for CommandResult, convert to `LocalRelation`. */
+  private def withCommandResultOptimized: Dataset[T] = {
+    logicalPlan match {
+      case c: CommandResult =>
+        // Convert to `LocalRelation` and let `ConvertToLocalRelation` do the casting locally to
+        // avoid triggering a job
+        Dataset(sparkSession, LocalRelation(c.output, c.rows))
```

Comment:
Can we do this in optimizer?

Reply:
This doesn't seem to be a universal optimization. Since the `getRows` and `isEmpty` methods only add a simple `Project` after the `CommandResult`, we don't need to trigger a job. However, if subsequent plans are more complex operations I think we could trigger a job.

Reply:
Maybe we should reconsider this: https://github.com/apache/spark/pull/40779#issuecomment-1506949427

Reply:
> Maybe we should reconsider this: [#40779 (comment)](https://github.com/apache/spark/pull/40779#issuecomment-1506949427)

Thanks for the tip, I will try it.

Reply:
Can I move `CommandResult.scala` into `spark-catalyst` module?

Reply:
> Maybe we should reconsider this: [#40779 (comment)](https://github.com/apache/spark/pull/40779#issuecomment-1506949427)

I submitted a new pr #45397 for this, could you please take a quick look?